core/modules/node/node.types.inc

```diff
@@ -683,11 +727,26 @@ function node_type_delete_confirm($form, &$form_state, $type) {
  * @see node_type_delete_confirm()
  */
 function node_type_delete_confirm_submit($form, &$form_state) {
-  $type_name = $form_state['values']['type'];
-  node_type_delete($type_name);
+  $type = node_type_load($form['type']['#value']);
+
+  $nodes = db_select('node', 'n')
+            ->fields('n', array('nid'))
+            ->condition('type', $type->type, '=')
+            ->execute();
+  foreach ($nodes as $node) {
+    $nids[] = $node->nid;
+  }
+
+  if (!empty($nids)) {
+    node_delete_multiple($nids);
```

Comment:
What worries me a bit here is the direct call to node_delete_multiple(), which potentially has to delete *a lot* of nodes and attached fields at once.

I'm not sure, but a batch job might make sense.